raiden/raiden_service.py

```diff
@@ -1001,9 +1042,15 @@ def _poll_until_target(self, target_block_number: BlockNumber) -> None:
 
             state_changes: List[StateChange] = list()
             for event in poll_result.events:
+                # Important: `blockchainevent_to_statechange` has to be called
+                # with the block of the current confirmed head! An unconfirmed
+                # block could lead to the wrong state being dispatched because
+                # of reorgs, and older blocks are not sufficient to fix
+                # problems with pruning, the `SyncTimeout` is used to ensure
+                # the `current_confirmed_head` stays valid.
                 state_changes.extend(
                     blockchainevent_to_statechange(
-                        self, event, poll_result.polled_block_number, pendingtokenregistration
+                        self, event, current_confirmed_head, pendingtokenregistration
```

Comment:
This was the bug, the value here was the confirmed block as seen by the state machine, which could be old enough to be pruned. The fix is to use the confirmed head instead.